drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1943 @@
+
+	k_mutex_unlock(&data->sync_mutex);
+}
+
+/* S/W connect */
+static inline void numaker_usbd_sw_connect(struct usb_dc_numaker_device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Clear all interrupts first for clean */
+	USBD_CLR_INT_FLAG(USBD_GET_INT_FLAG());
+
+	/* Enable relevant interrupts */
+	USBD_ENABLE_INT(USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET | USBD_INT_WAKEUP |
+			USBD_INT_SOF);
+
+	/* Clear SE0 for connect */
+	USBD_CLR_SE0();
+}
+
+/* S/W disconnect */
+static inline void numaker_usbd_sw_disconnect(struct usb_dc_numaker_device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Set SE0 for disconnect */
+	USBD_SET_SE0();
+}
+
+/* S/W disconnect, then connect*/
+static inline void numaker_usbd_sw_reconnect(struct usb_dc_numaker_device *dev)
+{
+	/* Keep SE0 to trigger bus reset */
+	numaker_usbd_sw_disconnect(dev);
+	k_sleep(K_USEC(NUMAKER_USBD_BUS_RESET_DRV_SE0_US));
+	numaker_usbd_sw_connect(dev);
+}
+
+/* Reset USB device address to 0 */
+static inline void numaker_usbd_reset_addr(struct usb_dc_numaker_device *dev)
+{
+	struct usb_dc_numaker_data *data = dev->data;
+
+	USBD_SET_ADDR(0);
+	data->addr = 0;
+}
+
+/* EP FIFO max count in bytes */
+static inline uint32_t numaker_usbd_ep_fifo_max(struct numaker_usbd_ep *ep_cur)
+{
+	/* NOTE: For one-shot implementation, effective size of EP FIFO is limited to EP MPS */
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+	__ASSERT_NO_MSG(ep_cur->ep_mps <= ep_cur->dmabuf_size);
+	return ep_cur->ep_mps;
+}
+
+/* EP FIFO used count in bytes */
+static inline uint32_t numaker_usbd_ep_fifo_used(struct numaker_usbd_ep *ep_cur)
+{
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+
+	return USB_EP_DIR_IS_OUT(ep_cur->ep_addr)
+		       ? ep_cur->read_fifo_used
+		       : numaker_usbd_ep_fifo_max(ep_cur) - ep_cur->write_fifo_free;
+}
+
+/* Set EP to stalled */
+static inline void numaker_usbd_ep_set_stall(struct numaker_usbd_ep *ep_cur)
+{
+	/* Set EP to stalled */
+	USBD_SET_EP_STALL(ep_cur->usbd_hw_ep_hndl);
+}
+
+/* Reset EP to unstalled and data toggle bit to 0 */
+static inline void numaker_usbd_ep_clear_stall_n_data_toggle(struct numaker_usbd_ep *ep_cur)
+{
+	/* Reset EP to unstalled */
+	USBD_CLR_EP_STALL(ep_cur->usbd_hw_ep_hndl);
+
+	/* Reset EP data toggle bit to 0 */
+	USBD_SET_DATA0(ep_cur->usbd_hw_ep_hndl);
+}
+
+/* Is EP stalled */
+static inline bool numaker_usbd_ep_is_stalled(struct numaker_usbd_ep *ep_cur)
+{
+	return USBD_GET_EP_STALL(ep_cur->usbd_hw_ep_hndl);
+}
+
+/* Send message */
+static int numaker_usbd_send_msg(const struct usb_dc_numaker_device *dev,
+				 const struct numaker_usbd_msg *msg)
+{
+	struct usb_dc_numaker_data *data = dev->data;
+	int rc;
+
+	rc = k_msgq_put(&data->msgq, msg, K_NO_WAIT);
+	if (rc < 0) {
+		LOG_ERR("Message queue overflow");
+
+		/* Discard all not yet received messages for error recovery below */
+		k_msgq_purge(&data->msgq);
+
+		/* Try to recover by S/W reconnect */
+		struct numaker_usbd_msg msg_reconn = {
```

Comment:
Variable definition should always be at the top of a block. Fix it everywhere.

Reply:
updated

Reply:
Not addressed.

Reply:
some missing. re-updated